control/state.py

```diff
@@ -160,159 +266,227 @@ def _remove_key(self, key: str):
+        self.local.remove_host(subsystem_nqn, host_nqn)
 
-        with rados.ReadOpCtx() as read_op:
-            iter, _ = self.ioctx.get_omap_vals(read_op, "", "", -1)
-            self.ioctx.operate_read_op(read_op, self.omap_name)
-            omap_dict = dict(iter)
-        return omap_dict
+    def add_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                     traddr: str, trsvcid: str, val: str):
+        """Adds a listener to the state data store."""
+        self.remote.add_listener(subsystem_nqn, gateway,
+                                 trtype, traddr, trsvcid, val)
+        self.local.add_listener(
+            subsystem_nqn, gateway, trtype, traddr, trsvcid, val)
+
+    def remove_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                        traddr: str, trsvcid: str):
+        """Removes a listener from the state data store."""
+        self.remote.remove_listener(
+            subsystem_nqn, gateway, trtype, traddr, trsvcid)
+        self.local.remove_listener(
+            subsystem_nqn, gateway, trtype, traddr, trsvcid)
 
     def delete_state(self):
-        """Deletes OMAP object."""
+        """Deletes state data stores."""
+        self.remote.delete_state()
+        self.local.delete_state()
+
+    def set_update_callbacks(self, gateway_rpcs):
+        self.gateway_rpcs = gateway_rpcs
+
+    def update(self,
+               notify_id=None,
+               notifier_id=None,
+               watch_id=None,
+               data=None):
+        """Checks for updated remote state and initiates local update."""
+
+        remote_version = self.remote.get_omap_version()
+        if int(self.remote.get_local_version()) < int(remote_version):
+            self.logger.debug("Outdated local state. Updating...")
+            local_state_dict = self.local.get_state()
+            remote_state_dict = self.remote.get_state()
+            # Handle OMAP additions
+            added_keys = remote_state_dict.keys() - local_state_dict.keys()
+            added = {key: remote_state_dict[key] for key in added_keys}
+            self._update_caller(added, True)
+            # Handle OMAP changes
+            same_keys = remote_state_dict.keys() & local_state_dict.keys()
+            changed = {
+                key: remote_state_dict[key]
+                for key in same_keys
+                if remote_state_dict[key] != local_state_dict[key]
+            }
+            self._update_caller(changed, True)
+            # Handle OMAP removals
+            removed_keys = local_state_dict.keys() - remote_state_dict.keys()
+            removed = {key: local_state_dict[key] for key in removed_keys}
+            self._update_caller(removed, False)
+            # Update local state and version
+            self.local.update(remote_state_dict)
+            self.remote.set_local_version(int(remote_version))
+            self.logger.debug("Update complete.")
```

Comment:
Wondering if the update would be so simple. E.g. I am not sure about how you handle `changed` keys. Suppose a bdev (rbd image) is changed. Actually, I suppose you can't just change image in bdev, you will need tor delete and create bdev, and for this you will need to recretead the namespace. So usually, when a "remote" gw is fast enough, it will just follow remove/add updates performed on the "local" gw. But suppose by some reason the "remote" gw missed some updates and now is catching up, so the bdev update is in `changed` now. Then it will try create_bdev, which will fail.

I might be not very clear, so let me know if you want me to elaborate. But my main thought is that I am not sure we may treat the `changed` the way you do it here. Probably we always need to delete/remove and then to create/add.

And the order of deleting may be different from the order of adding. E.g. we need to add a bdev before adding the namespace, but we can't remove the bdev before removing the namespace.